src/librustdoc/formats/item_type.rs

```diff
@@ -176,8 +171,10 @@ impl ItemType {
             DefKind::Variant => Self::Variant,
             DefKind::Field => Self::StructField,
             DefKind::AssocTy => Self::AssocType,
-            DefKind::AssocFn if let Some(DefKind::Trait) = parent_kind => Self::TyMethod,
-            DefKind::AssocFn => Self::Method,
+            DefKind::AssocFn if tcx.associated_item(def_id).defaultness(tcx).has_value() => {
+                Self::Method
+            }
+            DefKind::AssocFn => Self::TyMethod,
```

Comment:
nit: I would personally use an if/else here instead of a match guard, but it's subjective.